### PR TITLE
MBS-7652 Support local emulators in test runner scheduling

### PR DESCRIPTION
**Pull request by eugene-krivobokov**
Example of usage:

```groovy
configurationsContainer.register("Local") {
        tryToReRunOnTargetBranch = false
        reportSkippedTests = false
        rerunFailedTests = true
        reportFlakyTests = false

        targetsContainer.register("api27") {
            deviceName = "API27"

            scheduling = SchedulingConfiguration().apply {
                quota = QuotaConfiguration().apply {
                    retryCount = 1
                    minimumSuccessCount = 1
                }
                reservation = StaticDeviceReservationConfiguration().apply {
                    device = LocalEmulator.device(27)  // <---------
                    count = 1
                }
            }
        }
    }
```

### CODE REVIEW

File: subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/reservation/client/local/LocalReservationClient.kt [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+    private val configurationName: String,
+    private val logger: CILogger
+) : ReservationClient {
+
+    private var state: State = State.Idling
+
+    override suspend fun claim(
+        reservations: Collection<Reservation.Data>,
+        serialsChannel: SendChannel<String>,
+        reservationDeployments: SendChannel<String>
```
Comment (eugene-krivobokov) [inline subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/reservation/client/local/LocalReservationClient.kt:37]: For local deployments it contains stub.
Perhaps, we will store deployment here as a local state.

File: subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/InstrumentationTestsAction.kt [hunk 2]
```diff
@@ -106,7 +103,7 @@ class InstrumentationTestsAction(
         instrumentationConfiguration = params.instrumentationConfiguration,
         reportsApi = reportsApi,
         logger = logger,
-        registry = params.registry
+        kubernetesRegistry = params.kubernetesRegistry
```
Comment (dsvoronin) [inline subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/InstrumentationTestsAction.kt]: it is dockerRegistry in fact

File: subprojects/android-test/test-app/build.gradle.kts [hunk 3]
```diff
@@ -133,6 +134,29 @@ extensions.getByType<GradleInstrumentationPluginConfiguration>().apply {
         "jobSlug" to "FunctionalTests"
     )
 
+    configurationsContainer.register("Local") {
```
Comment (dsvoronin) [inline subprojects/android-test/test-app/build.gradle.kts]: it will be better to add this config simultaneously with version bump probably

**Review by dsvoronin — APPROVED** [review]

**Review by int02h — APPROVED** [review]
I think it will be much better if such large PRs would be divided into a few smaller PRs.

File: subprojects/android-test/test-app/build.gradle.kts (hunk 3 above)
Comment (eugene-krivobokov) [reply subprojects/android-test/test-app/build.gradle.kts]: removed

File: subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/InstrumentationTestsAction.kt (hunk 2 above)
Comment (eugene-krivobokov) [reply subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/InstrumentationTestsAction.kt]: thanks, fixed

File: subprojects/android-test/test-app/build.gradle.kts [hunk 4]
```diff
@@ -3,6 +3,7 @@ import com.avito.instrumentation.configuration.InstrumentationPluginConfiguratio
 import com.avito.instrumentation.configuration.target.scheduling.SchedulingConfiguration
 import com.avito.instrumentation.configuration.target.scheduling.quota.QuotaConfiguration
 import com.avito.instrumentation.configuration.target.scheduling.reservation.TestsBasedDevicesReservationConfiguration
+import com.avito.instrumentation.configuration.target.scheduling.reservation.StaticDeviceReservationConfiguration
```
Comment (sergioboi) [inline subprojects/android-test/test-app/build.gradle.kts]: unused?

File: subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/reservation/adb/AndroidDebugBridge.kt [hunk 5]
```diff
@@ -1,11 +1,25 @@
...
+        } else {
+            LocalDevice(
+                serial = serial,
+                logger = logger
+            )
+        }
+    }
+
+    private fun isRemote(serial: String): Boolean {
+        return serial.contains(':') && serial.substringBefore(':').contains('.')
```
Comment (sergioboi) [inline subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/reservation/adb/AndroidDebugBridge.kt]: Can we rely on something more stable and clear?

File: subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/reservation/adb/Device.kt [hunk 6]
```diff
@@ -41,12 +42,12 @@ class Device(
         errorMessage = "failed to connect to $serial"
     )
 
-    fun disconnect(): Try<String> = runCommand(
+    open fun disconnect(): Try<String> = runCommand(
```
Comment (sergioboi) [inline subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/reservation/adb/Device.kt:45]: Better is make abstract base class for default device actions

File: subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/reservation/client/local/LocalReservationClient.kt [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+                delay(TimeUnit.SECONDS.toMillis(5))
+
+                emulators = localEmulators(reservation)
+            }
+        }
+
+        return result
+    }
+
+    private fun generateDeploymentName(): String = "local-${UUID.randomUUID()}"
```
Comment (sergioboi) [inline subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/reservation/client/local/LocalReservationClient.kt:145]: `deployment` made harder to understand real function usage

**Comment by eugene-krivobokov** [comment]
> I think it will be much better if such large PRs would be divided into a few smaller PRs.

Agree. I've reverted some changes to reduce a scope.
The really new classes are only `LocalReservationClient` and `LocalDevice`. Other logic hasn't changed.

**Review by sergioboi — CHANGES_REQUESTED** [review]
Sorry but I don't understand who and where decides that we need a local run

File: subprojects/android-test/test-app/build.gradle.kts (hunk 4 above)
Comment (eugene-krivobokov) [reply subprojects/android-test/test-app/build.gradle.kts]: fixed

File: subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/reservation/adb/AndroidDebugBridge.kt (hunk 5 above)
Comment (eugene-krivobokov) [reply subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/reservation/adb/AndroidDebugBridge.kt]: The root of the problem is that we parse and rely on adb command output.
Here must be an IP address.
https://developer.android.com/studio/command-line/adb#wireless

I've changed to more correct check.

Comment (sergioboi) [reply subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/reservation/adb/AndroidDebugBridge.kt]: My idea is we need to separate serial type from device location. Let's discuss it in zoom

File: subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/reservation/client/ReservationClientFactory.kt [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+    ) : ReservationClientFactory {
+
+        override fun create(
+            configuration: InstrumentationConfiguration.Data,
+            executionParameters: ExecutionParameters,
+            testsToRun: List<TestWithTarget>
+        ): ReservationClient {
+            val hasLocal = testsToRun.any { it.target.reservation.device is Device.LocalEmulator }
+            val hasKubernetes = testsToRun.any { it.target.reservation.device is Device.Emulator }
+            if (hasLocal && hasKubernetes) {
```
Comment (sergioboi) [inline subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/reservation/client/ReservationClientFactory.kt]: We need to fail in configuration phase

**Review by sergioboi — CHANGES_REQUESTED** [review]
The whole idea is add a new type of device. It is quite simple.
But it is difficult to understand how to use it.

File: subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/reservation/client/ReservationClientFactory.kt (hunk 8 above)
Comment (eugene-krivobokov) [reply subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/reservation/client/ReservationClientFactory.kt]: added this check to a configuration phase

File: subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/configuration/InstrumentationConfiguration.kt [hunk 9]
```diff
@@ -104,6 +105,14 @@ open class InstrumentationConfiguration(val name: String) {
         val performanceType: PerformanceType?
     ) : Serializable {
 
+        init {
+            val hasLocal = targets.any { it.reservation.device is Device.LocalEmulator }
```
Comment (sergioboi) [inline subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/configuration/InstrumentationConfiguration.kt:109]: Are you sure we had all the targets registered at that moment?

File: subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/reservation/client/ReservationClientFactory.kt [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+                    logger = logger,
+                    buildId = buildId,
+                    buildType = buildType,
+                    emulatorsLogsReporter = emulatorsLogsReporter,
+                    registry = registry
+                )
+            }
+        }
+
+        private fun isLocalRun(testsToRun: List<TestWithTarget>): Boolean {
```
Comment (sergioboi) [inline subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/reservation/client/ReservationClientFactory.kt:84]: Duplication is an indicator of poor architecture. Add comment for future rewriting

File: subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/configuration/InstrumentationConfiguration.kt (hunk 9 above)
Comment (eugene-krivobokov) [reply subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/configuration/InstrumentationConfiguration.kt:109]: Yes, we build this DTO will all data. In any case it must have a valid state.

File: subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/reservation/client/ReservationClientFactory.kt (hunk 10 above)
Comment (eugene-krivobokov) [reply subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/reservation/client/ReservationClientFactory.kt:84]: > Duplication is an indicator of poor architecture

It depends.

File: subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/reservation/adb/Device.kt (hunk 6 above)
Comment (eugene-krivobokov) [reply subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/reservation/adb/Device.kt:45]: In the next PR

File: subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/reservation/client/ReservationClientFactory.kt (hunk 10 above)
Comment (eugene-krivobokov) [reply subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/reservation/client/ReservationClientFactory.kt:84]: done

File: subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/reservation/client/local/LocalReservationClient.kt (hunk 1 above)
Comment (eugene-krivobokov) [reply subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/reservation/client/local/LocalReservationClient.kt:37]: I will try to remove this knowledge from loca reservation later

**Comment by eugene-krivobokov** [comment]
> The whole idea is add a new type of device. It is quite simple.
> But it is difficult to understand how to use it.

Added an example to the PR's description

**Review by sergioboi — APPROVED** [review]

**Pull request merged**